practice-tool/src/widgets/character_stats.rs

```diff
@@ -5,15 +5,17 @@ use practice_tool_core::widgets::Widget;
 
 #[derive(Debug)]
 struct CharacterStatsEdit {
-    ptr: PointerChain<CharacterStats>,
+    stats_ptr: PointerChain<CharacterStats>,
+    blessings_ptr: Option<PointerChain<CharacterBlessings>>,
     stats: Option<CharacterStats>,
+    blessings: Option<CharacterBlessings>,
 }
 
 impl Stats for CharacterStatsEdit {
     fn data(&mut self) -> Option<impl Iterator<Item = Datum>> {
         self.stats.as_mut().map(|s| {
-            [
-                Datum::int("Level", &mut s.level, 1, i32::MAX),
+            let mut stats_data = vec![
```

Comment:
This is on the hot path, I'd prefer avoiding allocations and using a buffer preallocated into the struct and `.drain(..)`ed, but let's leave it like this for now. I need to make a pass over the whole codebase to ensure no allocations are made per frame anyway so I can deal with that myself.

Reply:
The allocations are only executed on each subwindow's popup.
Also, you can just make 2 pre-allocated vectors as caches, if time-save for the allocations are really needed

Reply:
IIUC it should be executed on every frame, as `imgui` is not retained. The input widgets are recreated at every frame and you can modify them while the popup is open.

So far per-frame allocations haven't been a huge deal honestly, but one of the design goals of the tool is to take as little frame time as possible to allow the game to maximise its usage for itself.